syncd/MdioIpcClient.cpp

```diff
@@ -0,0 +1,251 @@
+        if (connect(sock, (struct sockaddr *) &saddr, sizeof(saddr)) < 0)
+        {
+            SWSS_LOG_ERROR("connect() :%s", strerror(errno));
+            close(sock);
+            sock = 0;
+            unlink(caddr.sun_path);
+            return errno;
+        }
+    }
+
+    len = strlen(cmd);
+    ret = send(sock, cmd, len, 0);
+    if (ret < (ssize_t)len)
+    {
+        SWSS_LOG_ERROR("send failed, ret=%ld, expected=%ld\n", ret, len);
+        close(sock);
+        sock = 0;
+        unlink(caddr.sun_path);
+        return -EIO;
+    }
+
+    ret = recv(sock, resp, SYNCD_IPC_BUFF_SIZE - 1, 0);
+    if (ret <= 0)
+    {
+        SWSS_LOG_ERROR("recv failed, ret=%ld\n", ret);
+        close(sock);
+        sock = 0;
+        unlink(caddr.sun_path);
+        return -EIO;
+    }
+
+    timeout = time(NULL) + MDIO_CLIENT_TIMEOUT;
+    return (int)strtol(resp, NULL, 0);
+}
+
+
+/* Function to read data from MDIO interface */
+extern "C" sai_status_t mdio_read(uint64_t platform_context, uint32_t mdio_addr, uint32_t reg_addr,
+        uint32_t number_of_registers, uint32_t *data)
+{
+    // SWSS_LOG_ENTER(); // disabled
+
+    int rc = SAI_STATUS_FAILURE;
+    char cmd[SYNCD_IPC_BUFF_SIZE], resp[SYNCD_IPC_BUFF_SIZE];
+
+    if (number_of_registers > 1)
+    {
+        SWSS_LOG_ERROR("Multiple register reads are not supported, num_of_registers: %d\n", number_of_registers);
+        return SAI_STATUS_FAILURE;
+    }
+
+    ipcMutex.lock();
+
+    sprintf(cmd, "mdio 0x%x 0x%x\n", mdio_addr, reg_addr);
+    rc = syncd_mdio_ipc_command(cmd, resp);
+    if (rc == 0)
+    {
+        *data = (uint32_t)strtoul(strchrnul(resp, ' ') + 1, NULL, 0);
+        rc = SAI_STATUS_SUCCESS;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("syncd_mdio_ipc_command returns : %d\n", rc);
+    }
+
+    ipcMutex.unlock();
+    return rc;
+}
+
+/* Function to write data to MDIO interface */
+extern "C" sai_status_t mdio_write(uint64_t platform_context, uint32_t mdio_addr, uint32_t reg_addr,
+        uint32_t number_of_registers, const uint32_t *data)
+{
+    // SWSS_LOG_ENTER(); // disabled
+
+    int rc = SAI_STATUS_FAILURE;
+    char cmd[SYNCD_IPC_BUFF_SIZE], resp[SYNCD_IPC_BUFF_SIZE];
+
+    if (number_of_registers > 1)
+    {
+        SWSS_LOG_ERROR("Multiple register reads are not supported, num_of_registers: %d\n", number_of_registers);
+        return SAI_STATUS_FAILURE;
+    }
+
+    ipcMutex.lock();
+
+    sprintf(cmd, "mdio 0x%x 0x%x 0x%x\n", mdio_addr, reg_addr, *data);
+    rc = syncd_mdio_ipc_command(cmd, resp);
+    if (rc == 0)
+    {
+        rc = SAI_STATUS_SUCCESS;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("syncd_mdio_ipc_command returns : %d\n", rc);
+    }
+
+    ipcMutex.unlock();
+    return rc;
+}
+
+/* Function to read data using clause 22 from MDIO interface */
+extern "C" sai_status_t mdio_read_cl22(uint64_t platform_context, uint32_t mdio_addr, uint32_t reg_addr,
```

Comment:
all those extern C functions should be declared in extern C in header file, and not explicitly here

Reply:
which they are, so this extern should be redundant here